sig-azure/0018-20180711-azure-availability-zones.md

```diff
@@ -0,0 +1,281 @@
+
+```go
+func (l *persistentVolumeLabel) Admit(a admission.Attributes) (err error) {
+    ...
+    if volume.Spec.AzureDisk != nil {
+        labels, err := l.findAzureDiskLabels(volume)
+        if err != nil {
+            return admission.NewForbidden(a, fmt.Errorf("error querying AzureDisk volume %s: %v", volume.Spec.AzureDisk.DiskName, err))
+        }
+        volumeLabels = labels
+    }
+    ...
+}
+```
+
+> Note: the PersistentVolumeLabel admission controller will be deprecated, and cloud-controller-manager is preferred after its GA (probably v1.13 or v1.14).
+
+### StorageClass
+
+Note that the above interfaces are only applied to AzureDisk persistent volumes, not StorageClass. For AzureDisk StorageClass, we should add a few new options for zone-aware and [topology-aware](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md) provisioning. The following three new options will be added in AzureDisk StorageClass:
+
+- `zoned`: indicates whether new disks are provisioned with AZ. Default is `true`.
+- `zone` and `zones`: indicates which zones should be used to provision new disks (zone-aware provisioning). Only can be set if `zoned` is not false and `allowedTopologies` is not set.
+- `allowedTopologies`: indicates which topologies are allowed for [topology-aware](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md) provisioning. Only can be set if `zoned` is not false and `zone`/`zones` are not set.
+
+An example of zone-aware provisioning storage class is:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  annotations:
+  labels:
+    kubernetes.io/cluster-service: "true"
+  name: managed-premium
+parameters:
+  kind: Managed
+  storageaccounttype: Premium_LRS
+  # only one of zone and zones are allowed
+  zone: "centralus-1"
+  # zones: "centralus-1,centralus-2,centralus-3"
+provisioner: kubernetes.io/azure-disk
+```
+
+Another example of topology-aware provisioning storage class is:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  annotations:
+  labels:
+    kubernetes.io/cluster-service: "true"
+  name: managed-premium
+parameters:
+  kind: Managed
+  storageaccounttype: Premium_LRS
+provisioner: kubernetes.io/azure-disk
+allowedTopologies:
+- matchLabelExpressions:
+  - key: failure-domain.beta.kubernetes.io/zone
+    values:
+    - centralus-1
+    - centralus-2
+```
+
+AzureDisk can only be created with one specific zone, so if multiple zones are specified in the storage class, then new disks will be provisioned with zone chosen arbitrarily among them.
```

Comment:
Enabling support for `DynamicProvisioningScheduling` introduces several scenarios where the zone to provision a disk should not be chosen arbitrarily and specifying multiple zones in storage class should lead to an error:

If `DynamicProvisioningScheduling` is enabled and `VolumeBindingMode: WaitForFirstConsumer` is specified in the storage class, zone of the disk should be set to the zone of the node passed to `Provision()`. Specifying `zone`/`zones` in storage class should be considered an error in this scenario.

If `DynamicProvisioningScheduling` is enabled and `VolumeBindingMode: WaitForFirstConsumer` is not specified in StorageClass, zone of disk should be chosen from `allowedTopologies` or `zones` depending on which is specified. Specifying both `allowedTopologies` and `zones` should lead to error.

If `DynamicProvisioningScheduling` is disabled then the zone maybe arbitrarily chosen as specified above by arbitrarily choosing from the zones specified in the storage class.